prediction/src/preprocess/improved_lung_segmentation.py

```diff
@@ -0,0 +1,444 @@
+    lungs_seeds = skimage.morphology.binary_closing(lungs_seeds, selem)
+
+    lungs_seeds = lungs_postprocessing(lungs_seeds)
+
+    return lungs_seeds, trachea
+
+
+def cumulation(lung):
+    """The cumulative x-position inside a lung.
+
+    Args:
+        lung: The segmented lung (left or right)
+
+    Returns:
+        lung: cumulative x-position
+        z_coords: z-coordinates in lung
+
+    """
+    z_coords, y_coords, x_coords = np.where(lung)
+    current_volume = 0
+    volume = np.count_nonzero(lung)
+    for coord in np.unique(x_coords):
+        current_volume += np.count_nonzero(lung[:, :, coord])
+        lung[:, :, coord][lung[:, :, coord] != 0] = current_volume / float(volume)
+    return lung, z_coords
+
+
+def ventricular_extraction(err_lung):
+    """ The extraction of the ventricle.
+
+    On all CT there is a ventricle. In order for the algorithm
+    doesn't define the ventricle as a segmentation error,
+    it is necessary to remove the error mask of the largest
+    size that corresponds to the ventricle.
+
+    Args:
+        err_lung (np.ndarray[bool]): the label of errors in the lung.
+
+    Returns:
+        np.ndarray[bool]: err_lung - the label of errors in the lung without the ventricle.
+
+    """
+    max_marker = -1
+    z_max = -1
+    lav, mar = scipy.ndimage.label(err_lung)
+    volumes = np.hstack([np.bincount(lav.flatten()), [-1]])
+
+    for i in range(1, mar + 1):
+        z, y, x = np.where(lav == i)
+        curr_z = z.max()
+        if curr_z > z_max:
+            curr_z = z_max
+            coords = (z, y, x)
+
+        if curr_z == z_max:
+            if volumes[i] > volumes[max_marker]:
+                # curr_z = z_max
+                coords = (z, y, x)
+
+        err_lung[coords] = 0
+    return err_lung
+
+
+def costal_surface(lung, z_coords, max_coor, combined):
+    """ The costal surface.
+
+    The convexity is determined by comparing the costal lung surface
+    in axial slices to the convex hull of this costal lung surface.
+
+    Args:
+        lung: cumulative x-position in lung
+        z_coords: z-coordinates in lung
+        max_coor: threshold for the сostal surface for the lung
+        combined (np.ndarray[bool]): segmented lung (left or right)
+
+    """
+    erroneus = np.zeros(lung.shape)
+    for coord in np.unique(z_coords):
+        if max_coor == .8:
+            ROI = (lung[coord] < max_coor) * (lung[coord] != 0)
+        else:
+            ROI = (lung[coord] > max_coor) * (lung[coord] != 0)
+
+        erroneus[coord] = skimage.morphology.convex_hull_object(ROI) - ROI
+
+    erroneus = erroneus * (1 - (combined != 0))
+    erroneus_erosion = scipy.ndimage.morphology.binary_erosion(erroneus, structure=skimage.morphology.ball(4.5))
+    erroneus_erosion = scipy.ndimage.morphology.binary_dilation(erroneus_erosion,
+                                                                structure=skimage.morphology.ball(4.5))
```

Comment:
where does `4.5` come from?

Reply:
Also, how about saving `structure` to a variable instead of calculating it twice?

Reply:
Thank you, good remark!
4.5 was taken experimentally, in https://www.ncbi.nlm.nih.gov/pubmed/19673192 was 7, but I decided that this is a lot